File: mpp/lab4/server/src/main/java/repository/ArbiterRepository.java
package repository;

import domain.Arbiter;
import domain.ScoreType;

import java.sql.*;

public class ArbiterRepository extends DatabaseRepository
        implements IArbiterRepository {
    public ArbiterRepository(Connection connection) throws RepositoryError {
        super(connection);
    }

    @Override
    public void createTable() throws RepositoryError {
        if (isTableCreated("Arbiters")) {
            return;
        }

        String command =
                "create table Arbiters (" +
                    "ArbiterId integer primary key autoincrement, " +
                    "ArbiterName nvarchar(255) not null unique, " +
                    "ArbiterPassword nvarchar(255) not null, " +
                    "ArbiterType integer not null" +
                ");";

        try {
            Statement statement = connection.createStatement();
            statement.execute(command);
            logger.info("Created table Arbiters");
            logger.info(command);
        } catch (SQLException e) {
            logger.error("Failed to create table Arbiters");
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public void add(Arbiter arbiter) throws RepositoryError {
        String command =
                "insert into Arbiters (" +
                    "ArbiterName, " +
                    "ArbiterPassword, " +
                    "ArbiterType" +
                ") values (" +
                    "?, " +
                    "?, " +
                    "?" +
                ");";

        try {
            PreparedStatement statement = connection.prepareStatement(command, Statement.RETURN_GENERATED_KEYS);
            statement.setString(1, arbiter.getName());
            statement.setString(2, arbiter.getPassword());
            statement.setInt(3, arbiter.getType().ordinal());
            statement.executeUpdate();

            ResultSet results = statement.getGeneratedKeys();
            if (results.next()) {
                arbiter.setId(results.getInt(1));
            } else {
                throw new RepositoryError("Failed to retrieve generated keys");
            }

            logger.info("Added arbiter");
            logger.info(command);
        } catch (RepositoryError | SQLException e) {
            logger.error("Failed to add arbiter");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public Arbiter findByNameAndPassword(String name, String password) throws RepositoryError {
        String command =
                "select " +
                    "ArbiterId, " +
                    "ArbiterType " +
                "from Arbiters where " +
                    "ArbiterName=? and " +
                    "ArbiterPassword=?" +
                ";";

        try {
            PreparedStatement statement = connection.prepareStatement(command);
            statement.setString(1, name);
            statement.setString(2, password);
            ResultSet results = statement.executeQuery();

            Arbiter arbiter;
            if (results.next()) {
                int id = results.getInt(1);
                ScoreType type = ScoreType.values()[results.getInt(2)];
                arbiter = new Arbiter(id, name, password, type);
            } else {
                throw new RepositoryError("Failed to retrieve arbiter");
            }

            logger.info("Retrieved arbiter");
            logger.info(command);

            return arbiter;
        } catch (RepositoryError | SQLException e) {
            logger.error("Failed to retrieve arbiter");
            logger.error(e);
            throw new RepositoryError(e);
        }
    }
}

File: mpp/lab5/repository/ArbiterRepository.cs
using lab5.domain;
using lab5.utils;
using System.Data.Common;
using log4net;

namespace lab5.repository
{
    public class ArbiterRepository : IArbiterRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("ParticipantRepository");
        
        private DbConnection Connection { get; }

        public ArbiterRepository(DbConnection connection)
        {
            Connection = connection;
            CreateTable();
        }

        private void CreateTable()
        {
            var command = Connection.CreateCommand(
                @"
                    create table if not exists Arbiters (
                        ArbiterId integer primary key autoincrement,
                        ArbiterName nvarchar(255) not null unique,
                        ArbiterPassword nvarchar(255) not null,
                        ArbiterType integer not null
                    );
                ");
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to create table", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to create table", e);
            }
            
            Logger.Info("Created table or already existing");
            Logger.Info(command.CommandText);
        }

        public void Add(Arbiter arbiter)
        {
            var command = Connection.CreateCommand(
                @"
                    insert into Arbiters(
                        ArbiterName,
                        ArbiterPassword,
                        ArbiterType
                    ) values (
                        @ArbiterName,
                        @ArbiterPassword,
                        @ArbiterType
                    );
                ");
            command.AddParameterWithValue("@ArbiterName", arbiter.Name);
            command.AddParameterWithValue("@ArbiterPassword", arbiter.Password);
            command.AddParameterWithValue("@ArbiterType", (int)arbiter.Type);
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to add arbiter", e);
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to add arbiter", e);
            }
            
            Logger.Info("Added arbiter");
            Logger.Info(command.CommandText);

            command = Connection.CreateCommand(
                @"
                    select
                        max(ArbiterId)
                    from Arbiters where
                        ArbiterName=@ArbiterName and
                        ArbiterPassword=@ArbiterPassword
                    ;
                ");

            command.AddParameterWithValue("@ArbiterName", arbiter.Name);
            command.AddParameterWithValue("@ArbiterPassword", arbiter.Password);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to retrieve arbiter id");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to retrieve arbiter id", e);
            }

            if (!reader.Read())
            {
                Logger.Error("Failed to retrieve arbiter id");
                Logger.Error(command.CommandText);
                throw new RepositoryError("Failed to retrieve arbiter id");
            }
            else
            {
                arbiter.Id = reader.GetInt32(0);
            }

            Logger.Info("Retrieved arbiter id");
            Logger.Info(command.CommandText);
        }

        public Arbiter FindByNameAndPassword(string name, string password)
        {
            var command = Connection.CreateCommand(
                @"
                    select
                        ArbiterId,
                        ArbiterType
                    from Arbiters where
                        ArbiterName=@ArbiterName and
                        ArbiterPassword=@ArbiterPassword
                    ;
                ");
            command.AddParameterWithValue("@ArbiterName", name);
            command.AddParameterWithValue("@ArbiterPassword", password);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to retrieve arbiter by name and password");
                Logger.Error(command.CommandText);
                throw new RepositoryError("Failed to retrieve arbiter by name and password", e);
            }
            
            if (!reader.Read())
            {
                Logger.Error("Failed to retrieve arbiter by name and password");
                Logger.Error(command.CommandText);
                throw new RepositoryError("Failed to retrieve arbiter by name and password");
            }
            
            Logger.Info("Retrieved arbiter by name and password");
            Logger.Info(command.CommandText);
            
            var id = reader.GetInt32(0);
            var type = (ScoreType)reader.GetInt32(1);
            return new Arbiter(id, name, password, type);

        }
    }
}
